src/components/Sessions/RequireSession.js

```diff
@@ -1,42 +1,124 @@
+const WARN_SESSION_TIMEOUT_FACTOR = 0.9; // once 90% of the session expiry time has passed, show a warning
 
+const RelativeTimeToExpiry = ({numSeconds, raw = false}) => {
+  // more than 24 hours -> don't bother
+  if (numSeconds >= 3600 * 24) return null;
+  const component = (
+    <FormattedRelativeTime value={numSeconds} numeric="auto" updateIntervalInSeconds={1} />
+  );
+  if (raw) return component;
+  return <>&nbsp;({component})</>;
+};
+
+const untilWarningMilliseconds = (expiryDate) => {
+  /*
+  Calculate the time until the session expiry warning time should be displayed.
+   */
+  return (expiryDate - new Date()) * WARN_SESSION_TIMEOUT_FACTOR;
+};
+
+const RequireSession = ({expired = false, expiryDate = null, children}) => {
+  const {baseUrl} = useContext(ConfigContext);
+  const [secondsToExpiry, setSecondsToExpiry] = useState(0);
+  const [warningAccepted, setWarningAccepted] = useState(false);
+  const [warningVisible, setWarningVisible] = useState(false);
 
-const RequireSession = ({ expired=false, children }) => {
-  if (!expired) {
+  useEffect(() => {
+    /*
+    Show the warning if the session has expired.
+     */
+    if (expiryDate) {
+      const ms = untilWarningMilliseconds(expiryDate);
+      const timeout = setTimeout(() => {
+        setWarningVisible(true);
+      }, ms);
+      return () => clearTimeout(timeout);
+    }
+  }, [expiryDate]);
+
+  useEffect(() => {
+    /*
+    Count down to the session expiry time.
+     */
+    const interval = setInterval(() => {
+      setSecondsToExpiry(Math.floor((expiryDate - new Date()) / 1000));
+    }, 1000);
+    return () => clearInterval(interval);
+  }, [expiryDate]);
+
+  if (expired) {
     return (
-      <>
-        {children}
-        {/*<SessionExpiry /> disabled for now until we figure out how to style this */}
-      </>
+      <Card
+        title={<FormattedMessage description="Session expired card title" defaultMessage="Your session has expired"/>}>
+        <ErrorMessage>
+          <FormattedMessage
+            description="Session expired error message"
+            defaultMessage="Your session has expired. Click <link>here</link> to restart."
+            values={{
+              link: chunks => <Link to="/" component={Anchor}>{chunks}</Link>,
+            }}
+          />
+        </ErrorMessage>
+      </Card>
     );
+  }
+
+  const onExtendClick = async () => {
+    setWarningAccepted(true);
+    await apiCall(`${baseUrl}ping`);
   };
 
   return (
-    <Card title={<FormattedMessage description="Session expired card title" defaultMessage="Your session has expired" />}>
-      <ErrorMessage>
-        <FormattedMessage
-          description="Session expired error message"
-          defaultMessage="Your session has expired. Click <link>here</link> to restart."
-          values={{
-            link: chunks => <Link to="/" component={Anchor}>{chunks}</Link>,
-          }}
-        />
-      </ErrorMessage>
-    </Card>
-  );
+    <>
+      <Modal
```

Comment:
Hovering over the close icon displays `[Object object]`, so something is going wrong with the translations/title props there.